src/main/kotlin/org/rust/ide/annotator/RsUnsafeExpressionAnnotator.kt

```diff
@@ -14,29 +16,33 @@ import org.rust.lang.core.types.ty.TyPointer
 import org.rust.lang.core.types.type
 import org.rust.lang.utils.RsDiagnostic
 import org.rust.lang.utils.addToHolder
+import org.rust.openapiext.isUnitTestMode
 
 class RsUnsafeExpressionAnnotator : RsAnnotatorBase() {
     override fun annotateInternal(element: PsiElement, holder: AnnotationHolder) {
         val visitor = object : RsVisitor() {
             override fun visitCallExpr(o: RsCallExpr) = checkCall(o, holder)
             override fun visitMethodCall(o: RsMethodCall) = checkMethodCall(o, holder)
-            override fun visitPath(o: RsPath) = checkPath(o, holder)
+            override fun visitPathExpr(o: RsPathExpr) = checkPathExpr(o, holder)
             override fun visitUnaryExpr(o: RsUnaryExpr) = checkUnary(o, holder)
         }
 
         element.accept(visitor)
     }
 
-    private fun annotateUnsafeCall(o: RsExpr, holder: AnnotationHolder) {
-        if (!o.isInUnsafeBlockOrFn(/* skip the expression itself*/ 1)) {
-            RsDiagnostic.UnsafeError(o, "Call to unsafe function requires unsafe function or block").addToHolder(holder)
+    private fun annotateUnsafeCall(expr: RsExpr, holder: AnnotationHolder) {
+        if (!expr.isInUnsafeBlockOrFn(/* skip the expression itself*/ 1)) {
+            RsDiagnostic.UnsafeError(expr, "Call to unsafe function requires unsafe function or block").addToHolder(holder)
         } else {
-            val refElement = if (o is RsCallExpr) o.expr else o
-            holder.createUnsafeAnnotation(refElement, "Call to unsafe function")
+            val textRange = (expr as? RsCallExpr)?.expr?.textRange
+                ?: (expr as? RsDotExpr)?.methodCall
+                    ?.let { TextRange(it.startOffset, it.typeArgumentList?.endOffset ?: it.identifier.endOffset) }
+                ?: expr.textRange
```

Comment:
Let's convert this into `when` expr - I think it'll significantly improve readability of this piece of code